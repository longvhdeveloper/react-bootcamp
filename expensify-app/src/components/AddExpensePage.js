import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpeneseForm'
import { addExpense } from '../actions/expenses'

export class AddExpensePage extends React.Component {
  addExpense = (expense) => {
    this.props.addExpense(expense)
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={(expense) => {
            this.addExpense(expense)
          }}
        />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addExpense: (expense) => dispatch(addExpense(expense)),
  }
}

export default connect(undefined, mapDispatchToProps)(AddExpensePage)
