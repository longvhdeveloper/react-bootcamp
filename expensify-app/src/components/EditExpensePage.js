import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpeneseForm'
import { editExpense, removeExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {
  editExpense = (updated) => {
    const id = this.props.expense.id
    this.props.editExpense(id, updated)
    this.props.history.push('/')
  }

  removeExpense = () => {
    const id = this.props.expense.id
    this.props.removeExpense({ id })
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.editExpense} />
        <button onClick={this.removeExpense}>Remove</button>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch, props) => {
  return {
    editExpense: (id, updated) => dispatch(editExpense(id, updated)),
    removeExpense: () => dispatch(removeExpense()),
  }
}

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => {
      return expense.id === props.match.params.id
    }),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
