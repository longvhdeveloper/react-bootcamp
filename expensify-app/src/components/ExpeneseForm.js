import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'

// const date = new Date()
const now = moment()
console.log(now.format('MMM Do, YYYY'))

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      description: props.expense ? props.expense.description : '',
      notes: props.expense ? props.expense.notes : '',
      amount: props.expense ? props.expense.amount / 100 : '',
      createAt: props.expense ? moment(props.expense.createAt) : moment(),
      calendarFocused: false,
      error: '',
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value
    this.setState(() => ({ description }))
  }

  onNoteChange = (e) => {
    const notes = e.target.value
    this.setState(() => ({ notes }))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }

  onDateChange = (createAt) => {
    if (createAt) {
      this.setState(() => ({ createAt }))
    }
  }

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onSubmit = (e) => {
    e.preventDefault()

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount.' }))
    } else {
      this.setState(() => ({ error: '' }))
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createAt: this.state.createAt.valueOf(),
        notes: this.state.notes,
      })
    }
  }

  render() {
    return (
      <div>
        {this.state.error && <p className="error">{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            value={this.state.description}
            type="text"
            placeholder="Description"
            onChange={this.onDescriptionChange}
          />
          <input
            value={this.state.amount}
            type="text"
            placeholder="Amount"
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            onChange={this.onNoteChange}
            placeholder="Notes for expense"
          >
            {this.state.notes}
          </textarea>
          <button>Save</button>
        </form>
      </div>
    )
  }
}
