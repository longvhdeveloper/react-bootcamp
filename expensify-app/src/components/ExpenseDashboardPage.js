import React from 'react'

const ExpenseDashboardPage = () => {
  return <div>This is from my dashboard page</div>
}

export default ExpenseDashboardPage
