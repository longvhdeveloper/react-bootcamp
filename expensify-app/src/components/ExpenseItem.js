import React from 'react'
import { Link } from 'react-router-dom'

export const ExpneseItem = ({ expense }) => {
  return (
    <tr>
      <td>{expense.description}</td>
      <td>{expense.notes}</td>
      <td>{expense.amount}</td>
      <td>{expense.createAt}</td>
      <td>
        <Link to={`/edit/${expense.id}`}>Edit</Link>
      </td>
    </tr>
  )
}

export default ExpneseItem
