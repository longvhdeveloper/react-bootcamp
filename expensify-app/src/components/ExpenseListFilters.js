import React from 'react'
import { connect } from 'react-redux'
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from '../actions/filters'

import { DateRangePicker } from 'react-dates'

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  }

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate)
    this.props.setEndDate(endDate)
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }))
  }

  onTextChange = (e) => {
    // console.log(props)
    this.props.setTextFilter(e.target.value)
  }

  onSortChange = (e) => {
    switch (e.target.value) {
      case 'amount':
        this.props.sortByAmount()
        break

      default:
        this.props.sortByDate()
        break
    }
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />
        <select onChange={this.onSortChange}>
          <option selected={this.props.filters.sortBy === 'date'} value="date">
            Date
          </option>
          <option
            selected={this.props.filters.sortBy === 'amount'}
            value="amount"
          >
            Amount
          </option>
        </select>

        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={false}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    )
  }
}

const mapStateToProps = (state) => {
  return {
    filters: state.filters,
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)
