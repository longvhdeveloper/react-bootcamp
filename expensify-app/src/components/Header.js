import React from 'react'
import { NavLink } from 'react-router-dom'

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <ul>
      <li>
        <NavLink activeClassName="is-active" to="/" exact={true}>
          Home page
        </NavLink>
      </li>
      <li>
        <NavLink activeClassName="is-active" to="/create">
          Create expense page
        </NavLink>
      </li>
      <li>
        <NavLink activeClassName="is-active" to="/help">
          Help page
        </NavLink>
      </li>
    </ul>
  </header>
)

export default Header
