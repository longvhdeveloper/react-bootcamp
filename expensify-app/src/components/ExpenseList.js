import React from 'react'
import { connect } from 'react-redux'
import ExpenseItem from './ExpenseItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
  <div>
    <h1>ExpenseList</h1>
    <table>
      <thead>
        <tr>
          <th>Description</th>
          <th>Notes</th>
          <th>Amount</th>
          <th>Create at</th>
          <th>Action</th>
        </tr>
      </thead>
      <tbody>
        {props.expenses.length === 0 ? (
          <tr>
            <td colSpan="5">No expenses</td>
          </tr>
        ) : (
          props.expenses.map((expenseItem, index) => {
            return <ExpenseItem key={index} expense={expenseItem} />
          })
        )}
      </tbody>
    </table>
  </div>
)

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  }
}

export default connect(mapStateToProps)(ExpenseList)
