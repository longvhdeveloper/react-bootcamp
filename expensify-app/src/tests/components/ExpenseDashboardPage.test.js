import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseDashboardPage } from '../../components/ExpenseDashboardPage'

test('should reander ExpenseDashBoardPage', () => {
  const wrapper = shallow(<ExpenseDashboardPage />)
  expect(wrapper).toMatchSnapshot()
})
