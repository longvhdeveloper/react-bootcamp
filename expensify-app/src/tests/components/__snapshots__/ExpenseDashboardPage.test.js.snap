// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should reander ExpenseDashBoardPage 1`] = `
<div>
  <Connect(ExpenseListFilters) />
  <Connect(ExpenseList) />
</div>
`;
